src/modules/chat/markdown.rs

```diff
@@ -0,0 +1,413 @@
+    type Error = anyhow::Error;
+
+    fn try_from(value: Tag<'a>) -> Result<Self, Self::Error> {
+        let mapped = match value {
+            Tag::List(start) => EntityKind::List(start),
+            Tag::CodeBlock(lang) => EntityKind::TelegramEntityKind(MessageEntityKind::Pre {
+                language: lang.map(|lang| lang.to_string()),
+            }),
+            Tag::Italic => EntityKind::TelegramEntityKind(MessageEntityKind::Italic),
+            Tag::Bold => EntityKind::TelegramEntityKind(MessageEntityKind::Bold),
+            Tag::Strikethrough => EntityKind::TelegramEntityKind(MessageEntityKind::Strikethrough),
+            Tag::Link(url) | Tag::Image(url) => {
+                EntityKind::TelegramEntityKind(MessageEntityKind::TextLink {
+                    url: url.parse().unwrap(),
+                })
+            }
+            _ => {
+                return Err(anyhow!("Unexpected tag: {:?}", value));
+            }
+        };
+        Ok(mapped)
+    }
+}
+
+#[derive(Debug)]
+struct Entity {
+    kind: EntityKind,
+    start: usize,
+}
+
+const PARAGRAPH_MARGIN: usize = 2;
+const LIST_ITEM_MARGIN: usize = 1;
+
+#[derive(Debug)]
+struct ParseState {
+    entity_stack: Vec<Entity>,
+    parsed_string: ParsedString,
+    utf16_offset: usize,
+    prev_block_margin: usize,
+}
+
+impl ParseState {
+    fn new() -> Self {
+        Self {
+            entity_stack: Vec::new(),
+            parsed_string: ParsedString::default(),
+            utf16_offset: 0,
+            prev_block_margin: 0,
+        }
+    }
+
+    fn close(self) -> ParsedString {
+        let Self {
+            mut parsed_string,
+            prev_block_margin,
+            ..
+        } = self;
+
+        // Trim the redundant trailing margins.
+        parsed_string
+            .content
+            .truncate(parsed_string.content.len() - prev_block_margin);
+
+        parsed_string
+    }
+
+    fn next_state(mut self, event: Event) -> Self {
+        match event {
+            Event::Start(tag) => self.start(tag),
+            Event::End(tag) => self.end(tag),
+            Event::Text(text) => self.text(text),
+            Event::Code(text) => self.code(text),
+            Event::Break => self.r#break(),
+        };
+        self
+    }
+
+    fn start(&mut self, tag: Tag) {
+        match tag {
+            Tag::Paragraph => {}
+            Tag::Heading(level) => {
+                self.push_str(&format!("{} ", "#".repeat(level as _)));
+            }
+            Tag::Item => {
+                let item_marker = self
+                    .entity_stack
+                    .last()
+                    .and_then(|entity| match entity.kind {
+                        EntityKind::List(Some(start)) => Some(format!("{}. ", start)),
+                        EntityKind::List(None) => Some("• ".to_owned()),
+                        _ => None,
+                    })
+                    .expect("Expected a list entity");
```

Comment:
These operations may panic when there are some internal errors in the underlying parser library. It's better to isolate them and let them only panic in debug assertion mode.